# Fit the model or algorithm and use it to make predictions on data
![](https://media.giphy.com/media/lXo8uSnIkaB9e/giphy.gif)

In [1]:
#basic imports
import numpy as np
import pandas as pd 
import matplotlib as mlt

In [2]:
# get the dataset
hd = pd.read_csv("https://raw.githubusercontent.com/ineelhere/Machine-Learning-and-Data-Science/master/scikit-learn/heart-disease.csv")
hd.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

### Let us try fitting the model to the data first

Just in case --
* `x` = features/feature variables/data
* `y` = labels/targets/target variables

We fit the model at line 17

In [3]:
# import the ensemble classifiers
from sklearn.ensemble import RandomForestClassifier

# setup a random seed
np.random.seed(42)

# create the data
x = hd.drop("target", axis=1)
y = hd["target"]

# split into test and train sets
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2)

# fit the model to the data = training the ML model
clf = RandomForestClassifier()
clf.fit(x_train, y_train)

# evaluate the abohve fitted RandomForestClassifier model = use the pattens the ML model has learnt above!
clf.score(x_test,y_test)

0.8524590163934426

### Now let's see what we can predict  from the ML model
 We have 2 ways to make predictions -
 * the `predict()` function
 * the `predict_proba()` function
![](https://media.giphy.com/media/3oxHQF6UnJ0jkiclYA/giphy.gif)

In [4]:
x_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
179   57    1   0       150   276    0        0      112      1      0.6   
228   59    1   3       170   288    0        0      159      0      0.2   
111   57    1   2       150   126    1        1      173      0      0.2   
246   56    0   0       134   409    0        0      150      1      1.9   
60    71    0   2       110   265    1        0      130      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
249   69    1   2       140   254    0        0      146      0      2.0   
104   50    1   2       129   196    0        1      163      0      0.0   
300   68    1   0       144   193    1        1      141      0      3.4   
193   60    1   0       145   282    0        0      142      1      2.8   
184   50    1   0       150   243    0        0      128      0      2.6   

     slope  ca  thal  
179      1   1     1  
228      1   0     3  
111      2   1     3  
246      1   2     3  
60       2   1     2  
..     ...  ..   ...  
249      1   3     3  
104      2   0     2  
300      1   2     3  
193      1   2     3  
184      1   0     3  

[61 rows x 13 columns]

In [5]:
# use a trained model to make predictions
clf.predict(x_test)

array([0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [6]:
np.array(y_test)

array([0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

Now, do you notice a connection?<br>
`Our predictions are in the same format as the test labels`. <br>So, here y_test is our GROUND TRUTH LABEL

#### The `predict()` function

In [7]:
# compare predictions to truth labels to evaluate the model
y_preds = clf.predict(x_test) #making predictions with our model
np.mean(y_preds==y_test) # comparing the predictions to the ground truth labels

0.8524590163934426

In [8]:
# now here is an observation
clf.score(x_test, y_test)

0.8524590163934426

So what we're calculating here is the accuracy or just comparing how well the model did with this

In [9]:
# another way
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_preds)

0.8524590163934426

#### The `predict_proba()` function
Returns probabilities of a classification label. Returns the probability of the label being true <br>
https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict_proba

In [11]:
clf.predict_proba(x_test)

array([[0.89, 0.11],
       [0.49, 0.51],
       [0.43, 0.57],
       [0.84, 0.16],
       [0.18, 0.82],
       [0.14, 0.86],
       [0.36, 0.64],
       [0.95, 0.05],
       [0.99, 0.01],
       [0.47, 0.53],
       [0.26, 0.74],
       [0.7 , 0.3 ],
       [0.11, 0.89],
       [0.95, 0.05],
       [0.03, 0.97],
       [0.02, 0.98],
       [0.01, 0.99],
       [0.84, 0.16],
       [0.95, 0.05],
       [0.98, 0.02],
       [0.51, 0.49],
       [0.89, 0.11],
       [0.38, 0.62],
       [0.29, 0.71],
       [0.26, 0.74],
       [0.34, 0.66],
       [0.2 , 0.8 ],
       [0.22, 0.78],
       [0.83, 0.17],
       [0.15, 0.85],
       [0.94, 0.06],
       [0.92, 0.08],
       [0.96, 0.04],
       [0.62, 0.38],
       [0.46, 0.54],
       [0.89, 0.11],
       [0.44, 0.56],
       [0.16, 0.84],
       [0.33, 0.67],
       [0.08, 0.92],
       [0.13, 0.87],
       [0.17, 0.83],
       [0.18, 0.82],
       [0.38, 0.62],
       [0.32, 0.68],
       [0.77, 0.23],
       [0.39, 0.61],
       [0.  ,

So, that was the prediction of a row of a data if it indicates a heart disease. Notice that the data sum up to 1. <br>
For example, `in array([[0.89, 0.11], [0.49, 0.51],......` 0.89+0.11 = 1.0 and so on.
![](https://media.giphy.com/media/WJZdOtltegO76/giphy.gif)